qtfred/src/ui/dialogs/ShipEditor/ShipCustomWarpDialog.cpp

```diff
@@ -0,0 +1,212 @@
+	} else {
+		this->setWindowTitle("Edit Warp-in Parameters");
+	}
+	updateUI(true);
+	// Resize the dialog to the minimum size
+	resize(QDialog::sizeHint());
+}
+
+ShipCustomWarpDialog::~ShipCustomWarpDialog() = default;
+void ShipCustomWarpDialog::closeEvent(QCloseEvent* e)
+{
+	if (_model->query_modified()) {
+		auto button = _viewport->dialogProvider->showButtonDialog(DialogType::Question,
+			"Changes detected",
+			"Do you want to keep your changes?",
+			{DialogButton::Yes, DialogButton::No, DialogButton::Cancel});
+
+		if (button == DialogButton::Cancel) {
+			e->ignore();
+			return;
+		}
+
+		if (button == DialogButton::Yes) {
+			accept();
+			return;
+		}
+		if (button == DialogButton::No) {
+			_model->reject();
+		}
+	}
+	QDialog::closeEvent(e);
+}
+void ShipCustomWarpDialog::updateUI(const bool firstrun)
+{
+	util::SignalBlockers blockers(this);
+	if (firstrun) {
+		for (int i = 0; i < Num_warp_types; ++i) {
+			SCP_string name = Warp_types[i];
+			if (name == "Hyperspace") {
+				name = "Star Wars";
+			}
+			ui->comboBoxType->addItem(name.c_str());
+		}
+		for (const auto& fi : Fireball_info) {
+			ui->comboBoxType->addItem(fi.unique_id);
+		}
+		ui->comboBoxType->setCurrentIndex(_model->getType());
+
+		ui->lineEditStartSound->setText(_model->getStartSound().c_str());
+		ui->lineEditEndSound->setText(_model->getEndSound().c_str());
+		if (!_model->departMode()) {
+			ui->doubleSpinBoxEngage->setEnabled(false);
+			ui->labelEngageTime->setEnabled(false);
+		} else {
+			ui->doubleSpinBoxEngage->setValue(_model->getEngageTime());
+		}
+
+		ui->doubleSpinBoxSpeed->setValue(_model->getSpeed());
+		ui->doubleSpinBoxTime->setValue(_model->getTime());
+
+		if (_model->departMode()) {
+			ui->labelExponent->setText(tr("Deceleration Exponent:"));
+		} else {
+			ui->labelExponent->setText(tr("Acceleration Exponent:"));
+		}
+		ui->doubleSpinBoxExponent->setValue(_model->getExponent());
+		ui->doubleSpinBoxRadius->setValue(_model->getRadius());
+		ui->lineEditAnim->setText(_model->getAnim().c_str());
+		if (_model->getSupercap()) {
+			ui->checkBoxSupercap->setChecked(true);
+		}
+		if (!_model->isPlayer() || !_model->departMode()) {
+			ui->labelPlayerSpeed->setEnabled(false);
+			ui->doubleSpinBoxPlayerSpeed->setEnabled(false);
+		} else {
+			ui->doubleSpinBoxPlayerSpeed->setValue(_model->getPlayerSpeed());
+		}
+	}
+
+	if (ui->comboBoxType->itemText(ui->comboBoxType->currentIndex()) == "Homeworld") {
+		ui->lineEditAnim->setEnabled(true);
+		ui->labelAnim->setEnabled(true);
+	} else {
+		ui->lineEditAnim->setEnabled(false);
+		ui->labelAnim->setEnabled(false);
+	}
+
+	if (ui->comboBoxType->itemText(ui->comboBoxType->currentIndex()) == "Star Wars") {
+		ui->doubleSpinBoxExponent->setEnabled(true);
+		ui->labelExponent->setEnabled(true);
+	} else {
+		ui->doubleSpinBoxExponent->setEnabled(false);
+		ui->labelExponent->setEnabled(false);
+	}
+}
+void ShipCustomWarpDialog::rejectHandler()
```

Comment:
This seems like a common enough pattern that it should be in the superclass or in a utility function somewhere, though maybe that's beyond the scope of this PR.